### ISSUE TITLE
systemd-networkd removes fwmark rules (created by iproute2 command) on restart

### ISSUE THREAD

**Issue by cattyhouse** [systemd/systemd#19106]
**systemd version the issue has been seen with**
 > 247.4-2

**Used distribution**
 > archlinux

**Linux kernel version used** (`uname -a`)

 > `5.11.8-arch1-1`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > rule `32765:  from all fwmark 0x2333 lookup 100` remains 

**Unexpected behaviour you saw**
 > rule `32765:  from all fwmark 0x2333 lookup 100` gone

**Steps to reproduce the problem**
````
user at brix in ~
$ sudo ip rule add fwmark 0x2333 table 100

user at brix in ~
$ ip rule show
0:      from all lookup local
32765:  from all fwmark 0x2333 lookup 100
32766:  from all lookup main
32767:  from all lookup default

user at brix in ~
$ sudo systemctl restart systemd-networkd

user at brix in ~
$ ip rule show
0:      from all lookup local
32766:  from all lookup main
32767:  from all lookup default
````
**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

**Comment by cattyhouse** [comment]
update:

if append `protocol kernel`, the rule will not be removed by systemd-networkd, e.g.

sudo ip rule add fwmark 0x2333 table 100 `protocol kernel`

**Comment by bsdice** [comment]
> if append `protocol kernel`, the rule will not be removed by systemd-networkd, e.g.
> sudo ip rule add fwmark 0x2333 table 100 `protocol kernel`

I got bit by this too while trying to keep a blackhole default route with high metric put, even if the interface goes away, even if I do something like restart networkd, so traffic never goes out (leaks) via normal default gateway, but only via a tunnel interface. Compare:

https://github.com/systemd/systemd/blob/main/src/network/networkd-routing-policy-rule.c#L718
https://github.com/systemd/systemd/blame/main/src/network/networkd-route.c#L846
https://github.com/systemd/systemd/blame/main/src/network/networkd-route.c#L880

You are correct, networkd will leave proto kernel alone, but this is a hack. If you do a networkctl reconfigure ...some-interface... it will even wipe BGP or OSPF routes I presume? So, pick a free number:

> $ cat /etc/iproute2/rt_protos
> 0       unspec
> 1       redirect
> 2       kernel
> 3       boot
> 4       static
> **5       networkd**
> 8       gated
> 9       ra
> 10      mrt
> 11      zebra
> 12      bird
> 13      dnrouted
> 14      xorp
> 15      ntk
> 16      dhcp
> 18      keepalived
> 42      babel
> 186     bgp
> 187     isis
> 188     ospf
> 189     rip
> 192     eigrp

I am not sure though about the side-effects of networkd using its own number and cleaning up only what it "owns" in rules and routes. Also philosophically speaking.

**Comment by cattyhouse** [comment]
Thanks for the dig, are you suggesting something like `proto 9999`?

**Comment by bsdice** [comment]
That won't work, because if you look, the code only doesn't touch RTPROT_KERNEL (number 2 in the list) rules and routes. It will wipe, with a few peculiar exceptions, everything else. So the code would have to be changed, and also whoever maintains the RT proto list would need to be notified of a new route protocol named "networkd".

**Comment by cattyhouse** [comment]
> That won't work, because if you look, the code only doesn't touch RTPROT_KERNEL (number 2 in the list) rules and routes. It will wipe, with a few peculiar exceptions, everything else. So the code would have to be changed, and also whoever maintains the RT proto list would need to be notified of a new route protocol named "networkd".

I see. so this behaviour was added 6 months ago, interesting.

Well, i spent a long time (few hours) to discover that rules with `proto kernel` remain safely.

(in short: i tried to use systemd-networkd's network.conf to add rules, and it add the rules to `proto static` by default, then i read some man pages about `proto`, and tried `proto kernel`, then luckily, problem solved  )

**Comment by yuwata** [comment]
Hi, could you test PR #19287?

**Comment by cattyhouse** [comment]
Hi, @yuwata , thanks for the PR in such a short time, i'm on archlinux aarch64 && i'm not good at compiling systemd, sorry that i can't test the PR at the moment.

**Comment by bsdice** [comment]
@yuwata I tested your change with v248, works here. Thank you so much! One bug less that I have. ;-)

I am using a firewall service like so to start early:
```
# /etc/systemd/system/firewall.service
[Unit]
Description=Firewall
Before=network-pre.target
Wants=network-pre.target
After=local-fs.target
DefaultDependencies=no

[Service]
ExecStart=/etc/firewall/firewall start
Type=oneshot

[Install]
WantedBy=network.target
```
that sets up routing like this
```
IPv4

0:      from all lookup local
32765:  from all fwmark 0x19 lookup 250
32766:  from all lookup main
32767:  from all lookup default

default via 192.168.254.2 dev tun0 
unreachable default metric 9999 

IPv6

0:      from all lookup local
32765:  from all fwmark 0x19 lookup 250
32766:  from all lookup main

default via fddd:cafe:feed:ffff::2 dev tun0 metric 1024 pref medium
unreachable default dev lo metric 9999 pref medium
```
/etc/systemd/networkd.conf now has
```
ManageForeignRoutingPolicyRules=no
ManageForeignRoutes=no
```

**Comment by cattyhouse** [comment]
@yuwata i have a question, i'm on systemd 247.4 and man `networkd.conf` shows there is the option `ManageForeignRoutes=` , but i did NOT set this option in my eth0 network config (the only interface i have on this hardware), which means it defaults to yes. 

but restarting systemd-networkd will NOT remove this rule i created with `ip` command : `ip route add local 0/0 dev lo table 100`, is that something unexpected?

**Comment by bsdice** [comment]
@cattyhouse although (to quote Poettering's canned responses) this is not a support forum, let me help you compile and test systemd on Arch.

1. Update system to latest: pacman -Syu
2. Install base-devel group and asp package: pacman -S asp base-devel
3. In case you have other packages missing, makepkg will later complain. Just install those using "pacman -S --asdeps ..."
4. The following is done as a normal user
5. Make a temporary directory in your home and change into it: cd ~ ; mkdir tmp ; cd tmp
5. Checkout package build files for systemd: asp update systemd ; asp checkout systemd
6. cd systemd/trunk
7. Copy patches from your home into this directory, example:
```cp ~username/{0003-Hidepid.patch,4fac9422df5d40ce828585805a764d80c461ffed.patch,719eb5e40cf86771a4488c89989f8128c5c43263.patch} .```
8. Hint: To get such patches, navigate to like https://github.com/systemd/systemd/pull/19287/commits/719eb5e40cf86771a4488c89989f8128c5c43263 and append ".patch" to the URL.
9. Edit PKGBUILD like so (I have the hidepid patch in there as well, ignore that)
```bash
--- PKGBUILD.ORIG    2021-04-12 01:06:55.856584709 +0200
+++ PKGBUILD      2021-04-11 19:31:03.695795564 +0200
@@ -9,6 +9,7 @@
 pkgver="${_tag_name/-/}"
 pkgrel=4
 arch=('x86_64')
+groups=('custom')
 url='https://www.github.com/systemd/systemd'
 makedepends=('acl' 'cryptsetup' 'docbook-xsl' 'gperf' 'lz4' 'xz' 'pam' 'libelf'
              'intltool' 'iptables' 'kmod' 'libcap' 'libidn2' 'libgcrypt'
@@ -23,6 +24,9 @@
         "git+https://github.com/systemd/systemd#tag=v${_tag_name%.*}?signed"
         '0001-Use-Arch-Linux-device-access-groups.patch'
         '0002-Disable-SYSTEMD_URLIFY-by-default.patch'
+        '0003-Hidepid.patch'
+               '4fac9422df5d40ce828585805a764d80c461ffed.patch'
+               '719eb5e40cf86771a4488c89989f8128c5c43263.patch'
         'initcpio-hook-udev'
         'initcpio-install-systemd'
         'initcpio-install-udev'
@@ -44,6 +48,9 @@
             'SKIP'
             '882e486b6d88c8bafc50088845e41a49686e98981967f72ca1fb4ef07a01767400632f4b648fd31857d2a2a24a8fd65bcc2a8983284dd4fff2380732741d4c41'
             '313f3d6cc3d88f718509007e029213a82d84b196afdadc6ef560580acf70ab480aaecd7622f51726cc1af7d7841c6ec5390f72890b055a54fc74722341395651'
+                       'SKIP'
+                       'SKIP'
+                       'SKIP'
             'f0d933e8c6064ed830dec54049b0a01e27be87203208f6ae982f10fb4eddc7258cb2919d594cbfb9a33e74c3510cfd682f3416ba8e804387ab87d1a217eb4b73'
             'f599e1a35cba2c4e83e37c2299fac23ae128d8f68081283e71e1729384975dee1c4b677787f31a17890aeb98c8d2fc90405a202644290708ef9c027315022b17'
             'a25b28af2e8c516c3a2eec4e64b8c7f70c21f974af4a955a4a9d45fd3e3ff0d2a98b4419fe425d47152d5acae77d64e69d8d014a7209524b75a81b0edb10bf3a'
@@ -89,6 +96,13 @@
 
   # https://github.com/gwsw/less/issues/140
   patch -Np1 -i ../0002-Disable-SYSTEMD_URLIFY-by-default.patch
+
+  # https://github.com/systemd/systemd/issues/16896
+  patch -Np1 -i ../0003-Hidepid.patch
+
+  # https://github.com/systemd/systemd/issues/19106
+  patch -Np1 -i ../4fac9422df5d40ce828585805a764d80c461ffed.patch
+  patch -Np1 -i ../719eb5e40cf86771a4488c89989f8128c5c43263.patch
 }
 
 build() {
```
10. makepkg -Ccf
11. As root, install relevant <.zst files> pacman -U ...zst. In my case that is systemd systemd-libs and systemd-sysvcompat, I leave out systemd-resolvconf because this machine has manually configured dnsmasq and unbound for DNS.
12. Edit /etc/pacman.conf for good measure to prevent fixed packages from being overwritten by pacman -Syu. After a while this becomes dangerous because incompatible libs might get installed. For continued productive operation I use findbrokenpkgs from AUR with a "known missing" /etc/findbrokenpkgs/arch.conf file [1] that I create using a script [2] with empty arch.conf lines SEARCH_DIRS_MASK="" and LD_LIBRARY_MASK="".
```
IgnoreGroup = custom
```
13. Edit /etc/systemd/networkd.conf
```
ManageForeignRoutingPolicyRules=no
ManageForeignRoutes=no
```
14. Reboot and check results

Hope this helps!

[1] /etc/findbrokenpkgs/arch.conf example
```
SEARCH_DIRS="/usr/bin /usr/lib /usr/local/bin /usr/local/sbin /usr/local/lib /opt"
SEARCH_DIRS_MASK="/usr/lib/ImageMagick-7.0.11/modules-Q16HDRI"
LD_LIBRARY_MASK="libao.so.4 libatm.so.1 libespeak-ng.so.1 libfbclient.so.2 libfido2.so.1 libgcr-base-3.so.1 libGLU.so.1 libglut.so.3 libKF5CoreAddons.so.5 libKF5I18n.so.5 libKF5Wallet.so.5 libmysofa.so.1 libodbc.so.2 libQt5Bluetooth.so.5 libQt5Designer.so.5 libQt5Help.so.5 libQt5Multimedia.so.5 libQt5MultimediaWidgets.so.5 libQt5NetworkAuth.so.5 libQt5Nfc.so.5 libQt5Quick3D.so.5 libQt5RemoteObjects.so.5 libQt5SerialPort.so.5 libQt5TextToSpeech.so.5 libQt5WebSockets.so.5 libQt5X11Extras.so.5 libQt5XmlPatterns.so.5 librrd.so.8 libR.so libruby.so.2.7 libsybdb.so.5 libtk8.6.so libtwolame.so.0 libxul.so libQt6Core.so.6 libQt6Gui.so.6 libQt6Widgets.so.6 libclucene-core.so.1 libclucene-shared.so.1 libaspell.so.15 libhspell.so.0 libnuspell.so.4 libvoikko.so.1"
```
[2] /usr/local/bin/collect-missing-libs
```awk
#!/bin/sh

findbrokenpkgs -nc -nw 2>&1 | \
        awk '
                BEGIN {
                        FS="needs missing"
                }
                / needs missing / {
                        split($2,a," ")
                        for (l in a) arr[a[l]]=1
                }
                END {
                        x=0
                        for (l in arr) {
                                if (x==1) printf " %s",l
                                else {
                                        x=1
                                        printf "%s",l
                                }
                        }
                        printf "\n"
                }
        ' | \
                xargs -n1 | sort -u | xargs
```

**Comment by jluebbe** [comment]
I stumbled across this when updating from Debian bullseye (systemd 247.3) to bookworm (systemd 252.12).

The system is use as a router and networkd is configured to manage only a few specific network devices, each with a separate routing table using `RouteTable=203` in the `[DHCP]` section. Other interfaces, policy routing (`ip rule`) and iptables are configured statically (using Debian's `ifupdown`).

After the update, most of the rules added via `ip rule add` were missing, although sometimes the one or two last rules survived. With `ip monitor` I then found that they were being deleted shortly after being added. Having networkd suddenly start modifying global (non-interface specific) state was very surprising. Setting `ManageForeignRoutingPolicyRules=no` fixed the issue.

While it's probably too late to change the behavior again, perhaps this makes it easier to google and helps others.
